contracts/SimpleGAMultiSig.aes

```diff
@@ -0,0 +1,153 @@
+// ISC License
+//
+// Copyright (c) 2022, aeternity developers
+//
+// Credit to Michel Meier & Marco Walz (kryptokrauts.com)
+//
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
+// REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
+// AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
+// INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
+// LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE
+// OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
+// PERFORMANCE OF THIS SOFTWARE.
+
+// THIS IS NOT SECURITY AUDITED
+// DO NEVER USE THIS WITHOUT SECURITY AUDIT FIRST
+
+@compiler >= 6
+
+include "Option.aes"
+include "String.aes"
+
+// This contract provides multi signature support based on generalized accounts
+// A transaction can be proposed by one of the multisig users which are registered during contract deployment
+// The transaction must be confirmed by state.confirmations_required of the registered multisig users, while the proposing user automatically confirms
+// If the amount of confirmations is reached, authenticate can be called by any of the multisig users and the original tx is processed
+// Currently only one proposed transaction can be multisigned at a time
+contract SimpleGAMultiSig =
+
+  datatype event
+    = TxProposed(hash, address, int)
+    | TxConfirmed(hash, address)
+    | TxConsensusReached(hash)
+    | TxAuthorized(hash)
+    | TxRevoked(address)
+
+  record transaction = 
+    { ga_tx_hash : hash,
+      expires_at : int,
+      proposed_by : address,
+      confirmed_by : list(address) }
+
+  record state = 
+  	{ multisig_users : list(address),
+  	  confirmations_required : int,    
+      current_tx : option(transaction),
+      nonce : int,
+      owner: address }
+
+  stateful entrypoint init(p_confirmations_required : int, p_multisig_users : list(address)) = 
+    { multisig_users = Call.caller :: p_multisig_users, 
+  	  confirmations_required = p_confirmations_required,
+  	  current_tx = None,
+  	  nonce = 1,
+      owner = Call.caller }
+
+  // GA authorize function
+  stateful entrypoint authorize(sig: signature, n : int) : bool =  
+    check_nonce(n)
```

Comment:
I'd rename to indicate that it does 1. check, 2. abort and 3. set a new nonce

Reply:
we can take whatever name we want to of course. but this is a special GA function used in the auth context of a GA. not sure if we need to change the name of this specific function.

but I just checked the logic and I think the signature check for the tx is not needed as the tx is already confirmed by enough parties to achieve consensus

Reply:
I am wondering if the nonce in this "one tx at a time" scenario even makes sense. I think we could completely drop it. what do you think?

Reply:
I dropped the nonce handling. I think it's irrelevant